### ISSUE TITLE
WCAG, JAWS, and HTML5 accessibility

### ISSUE THREAD

**Issue by mktung** [Submitty/Submitty#4491]
Following John Fisher's accessibility overview, there are a number of things we need to fix. I've pushed solutinos to some of them in #4489 and these are the rest:

- [x] Script between head and body
- [ ] JS mimicking link
- [ ] Link text replaced by aria-label attribute
- [x] Element meta missing itemprop
- [x] Input descendants of label must have matching id to for
- [ ] Each \<a> element must contain text or an img with an alt

Forum Specific:
- [x] Hidden h1
- [x] Empty headings
- [x] Duplicate ID
- [x] Onclick should also have keyboard option
- [x] Detached labels
- [x] Input cannot be descendant of \<a>
- [ ] \<input type=text> with aria label

**Comment by mktung** [comment]
One thing I'm concerned about is:
- Element meta missing itemprop

He says this is an HTML5 spec but I can't find a source for it. I'll keep it in the issue but I don't have plans to fix this unless it's explained to me. From what I've read on MDN, W3C, and SO, it seems to imply that either you use the name or itemprop tag but not both (some suggest having 2 meta tags to capture both options). This is contradictory to the issue that we should add it.

**Comment by bmcutler** [comment]
Tagging @pbailie @mcglow2

**Comment by elihschiff** [comment]
Do you have more details on what the problems I listed below mean, or where you got them from (ex: WAVE, validator.w3.org, etc.) so I can understand the problems and how to fix them:

"JS mimicking link"
"Link text replaced by aria-label attribute"
"Detached labels"
"\<input type=text> with aria label"

**Comment by pbailie** [comment]
According to the report:
* "JS mimicking link" is in regards to WCAG 2.0 A F42.
* "Link text replaced by aria-label attribute" is an issue related to JAWS 2018 assistive technology, and maybe others.
* "Detached labels" is in regards to WCAG 2.0 A 1.4.2.  Maybe there are form labels that are not properly associated with form control IDs.
* "<input type=text> with aria label" is an issue with the NVDA 2018 assistive technology, and possibly others.

As an aside:  John Fisher told me that Rensselaer requires WCAG AA conformance, although AAA conformance is encouraged.

**Comment by elihschiff** [comment]
I looked and cannot find any examples of "JS mimicking link", is there a way to use a tool to detect this? I was searching with regex based off what this said the issue is https://www.w3.org/TR/WCAG20-TECHS/F42.html

Also is there a way to detect "Each <a> element must contain text or an img with an alt" I am having trouble coming up with a way to find locations where this happens. 

In order to solve "Link text replaced by aria-label attribute" we cannot have `<a aria-label=""></a>` However we seem to only do this when there is an icon inside the link but no text with that icon. Would a valid fix be to move the aria-label from the \<a> tag to the icon tag inside inside it?
so from this `<a aria-label=""><i></i></a>` to `<a><i aria-label=""></i></a>`

I am not sure how to solve "<input type=text> with aria label".  I assume the solution is not just to remove the aria-labels. Can I switch all the aria-label to aria-labelledby?

For the "Hidden h1" the only way to solve this is to unhide it ¯\\\_(ツ)\_/¯ but I have been told this has been discussed before so while I am happy to make the code unhide it I do not know if that is something that everyone agrees with.

Everything else in this issue has been solved already or I have fixed in my own branch and will be making a pr for soon.

**Comment by pbailie** [comment]
> I looked and cannot find any examples of "JS mimicking link", is there a way to use a tool to detect this? I was searching with regex based off what this said the issue is https://www.w3.org/TR/WCAG20-TECHS/F42.html

John Fisher's report was done last Sept.  It's possible this was fixed in some other work.  I'll try to vet this one again.

> Also is there a way to detect "Each element must contain text or an img with an alt" I am having trouble coming up with a way to find locations where this happens.

WAVE will spot an `img` without an `alt` property.  The W3C validator will spot elements without text that are not allowed to be so by HTML 5 spec.  But in the report , this is specific to `<a>` elements.

I think a lot of these involve font icons, which are links with neither text or images.  Font icons are part of a larger ongoing discussion.  

>In order to solve "Link text replaced by aria-label attribute" we cannot have <a aria-label=""></a> However we seem to only do this when there is an icon inside the link but no text with that icon. Would a valid fix be to move the aria-label from the <a> tag to the icon tag inside inside it?
>so from this <a aria-label=""><i></i></a> to <a><i aria-label=""></i></a>

Font icons... Stay tuned.

> I am not sure how to solve "<input type=text> with aria label". I assume the solution is not just to remove the aria-labels. Can I switch all the aria-label to aria-labelledby?

This problem is specific to NVDA 2018 screen reader, and maybe others.  There may need to be some more internal discussion over this.

> For the "Hidden h1" the only way to solve this is to unhide it ¯\_(ツ)_/¯ but I have been told this has been discussed before so while I am happy to make the code unhide it I do not know if that is something that everyone agrees with.

Last I had heard, @bmcutler does not want this changed.  I opened #5038 for this because John Fisher's report really highlighted this one (and it was worked on and closed?).  The resolution needs to be discussed on a higher level.

> Everything else in this issue has been solved already or I have fixed in my own branch and will be making a pr for soon.

Awesome!  Thank you for helping!

**Comment by bmcutler** [comment]
Correction:  it's not that I don't want the "hidden h1" on the forum page changed, it's that we should prioritize fixing more significant VPAT errors first.

@elihschiff  if you want to tackle the hidden h1 problem on the forum page now, please do.  You're already working on the breadcrumb code.  

Please make the entire breadcrumb on the forum page the h1 element for that page.  Use CSS to style the font so that it matches the breadcrumbs on the other page.  Every page except the forum has room for a full and natural h1 header (duplicating the text from the breadcrumb), so the breadcrumb is just text on the page, not an h tag.  Yes, it will mean adding some if logic to special case things.  We can talk on Wednesday if this design is unclear or doesn't work in the code for some reason.

**Comment by elihschiff** [comment]
~~it's that we should prioritize fixing more significant VPAT errors first~~
Too late :eyes:

**Comment by bmcutler** [comment]
You *have* been on a tearing through fixing the other VPAT issues :)
Thanks for doing the breadcrumb thing.  I hope it is the acceptable solution.

**Comment by bmcutler** [comment]
@elihschiff 
Do you think you've addressed everything in this issue?  Can you please update the checkboxes and/or close the issue?

**Comment by elihschiff** [comment]
I think it is probably safe to just close the issue at this point. Moat if not all of it has been addressed
